<!-- -->
<launch>

    <!-- op_common_params settings -->
    <arg name="pathDensity" default="/adeye/motion_planning/op_common_params/pathDensity"/>
    <arg name="rollOutDensity" default="/adeye/motion_planning/op_common_params/rollOutDensity"/>
    <arg name="rollOutsNumber" default="/adeye/motion_planning/op_common_params/rollOutsNumber"/>

    <arg name="maxVelocity" default="/adeye/motion_planning/op_common_params/maxVelocity"/>
    <arg name="minVelocity" default="/adeye/motion_planning/op_common_params/minVelocity"/>
    <arg name="maxLocalPlanDistance" default="/adeye/motion_planning/op_common_params/maxLocalPathDistance"/>
    <arg name="horizonDistance" default="/adeye/motion_planning/op_common_params/horizonDistance"/>

    <arg name="minFollowingDistance" default="/adeye/motion_planning/op_common_params/minFollowingDistance"/> <!-- should be bigger than Distance to follow -->
    <arg name="minDistanceToAvoid"
         default="/adeye/motion_planning/op_common_params/minDistanceToAvoid"/> <!-- should be smaller than minFollowingDistance and larger than maxDistanceToAvoid -->
    <arg name="maxDistanceToAvoid" default="/adeye/motion_planning/op_common_params/maxDistanceToAvoid"/> <!-- should be smaller than minDistanceToAvoid -->
    <arg name="speedProfileFactor" default="/adeye/motion_planning/op_common_params/speedProfileFactor"/>

    <arg name="smoothingDataWeight" default="/adeye/motion_planning/op_common_params/smoothingDataWeight"/>
    <arg name="smoothingSmoothWeight" default="/adeye/motion_planning/op_common_params/smoothingSmoothWeight"/>

    <arg name="horizontalSafetyDistance" default="/adeye/motion_planning/op_common_params/horizontalSafetyDistance"/>
    <arg name="verticalSafetyDistance" default="/adeye/motion_planning/op_common_params/verticalSafetyDistance"/>

    <arg name="enableSwerving" default="/adeye/motion_planning/op_common_params/enableSwerving"/>
    <arg name="enableFollowing" default="/adeye/motion_planning/op_common_params/enableFollowing"/>
    <arg name="enableTrafficLightBehavior" default="/adeye/motion_planning/op_common_params/enableTrafficLightBehavior"/>
    <arg name="enableStopSignBehavior" default="/adeye/motion_planning/op_common_params/enableStopSignBehavior"/>
    <arg name="enableLaneChange" default="/adeye/motion_planning/op_common_params/enableLaneChange"/>

    <arg name="width" default="/adeye/motion_planning/op_common_params/width"/>
    <arg name="length" default="/adeye/motion_planning/op_common_params/length"/>
    <arg name="wheelBaseLength" default="/adeye/motion_planning/op_common_params/wheelBaseLength"/>
    <arg name="turningRadius" default="/adeye/motion_planning/op_common_params/turningRadius"/> <!-- The original value was 5.2-->
    <arg name="maxSteerAngle" default="/adeye/motion_planning/op_common_params/maxSteerAngle"/> <!-- The original value was 0.45-->

    <arg name="steeringDelay" default="/adeye/motion_planning/op_common_params/SteeringDelay"/>
    <arg name="minPursuiteDistance" default="/adeye/motion_planning/op_common_params/minPursuitDistance"/>
    <arg name="additionalBrakingDistance" default="/adeye/motion_planning/op_common_params/additionalBrakingDistance"/>

    <arg name="giveUpDistance" default="/adeye/motion_planning/op_common_params/giveUpDistance"/>

    <arg name="maxAcceleration" default="/adeye/motion_planning/op_common_params/maxAcceleration"/>
    <arg name="maxDeceleration" default="/adeye/motion_planning/op_common_params/maxDeceleration"/>

    <arg name="velocitySource"
         default="/adeye/motion_planning/op_common_params/velocitySource"/> <!-- read velocities from (0- Odometry, 1- autoware current_velocities, 2- car_info) "" -->

    <!--waypoint_follower pure_pursuit settings-->
    <arg name="is_linear_interpolation" default="/adeye/motion_planning/pure_pursuit/is_linear_interpolation"/>
    <arg name="publishes_for_steering_robot" default="/adeye/motion_planning/pure_pursuit/publishes_for_steering_robot"/>

    <!-- Link to Simulink -->
    <include file="$(find adeye)/launch/sender.launch"/>

    <!-- op_common_params -->
    <include file="$(find op_local_planner)/launch/op_common_params.launch">
        <arg name="pathDensity" value="$(arg pathDensity)"/>
        <arg name="rollOutDensity" value="$(arg rollOutDensity)"/>
        <arg name="rollOutsNumber" value="$(arg rollOutsNumber)"/>

        <arg name="maxVelocity" value="$(arg maxVelocity)"/>
        <arg name="minVelocity" value="$(arg minVelocity)"/>
        <arg name="maxLocalPlanDistance" value="$(arg maxLocalPlanDistance)"/>
        <arg name="horizonDistance" value="$(arg horizonDistance)"/>

        <arg name="minFollowingDistance" value="$(arg minFollowingDistance)"/>
        <arg name="minDistanceToAvoid" value="$(arg minDistanceToAvoid)"/>
        <arg name="maxDistanceToAvoid" value="$(arg maxDistanceToAvoid)"/>
        <arg name="speedProfileFactor" value="$(arg speedProfileFactor)"/>

        <arg name="smoothingDataWeight" value="$(arg smoothingDataWeight)"/>
        <arg name="smoothingSmoothWeight" value="$(arg smoothingSmoothWeight)"/>

        <arg name="horizontalSafetyDistance" value="$(arg horizontalSafetyDistance)"/>
        <arg name="verticalSafetyDistance" value="$(arg verticalSafetyDistance)"/>

        <arg name="enableSwerving" value="$(arg enableSwerving)"/>
        <arg name="enableFollowing" value="$(arg enableFollowing)"/>
        <arg name="enableTrafficLightBehavior" value="$(arg enableTrafficLightBehavior)"/>
        <arg name="enableStopSignBehavior" value="$(arg enableStopSignBehavior)"/>
        <arg name="enableLaneChange" value="$(arg enableLaneChange)"/>

        <arg name="width" value="$(arg width)"/>
        <arg name="length" value="$(arg length)"/>
        <arg name="wheelBaseLength" value="$(arg wheelBaseLength)"/>
        <arg name="turningRadius" value="$(arg turningRadius)"/>
        <arg name="maxSteerAngle" value="$(arg maxSteerAngle)"/>

        <arg name="steeringDelay" value="$(arg steeringDelay)"/>
        <arg name="minPursuiteDistance" value="$(arg minPursuiteDistance)"/>
        <arg name="additionalBrakingDistance" value="$(arg additionalBrakingDistance)"/>

        <arg name="giveUpDistance" value="$(arg giveUpDistance)"/>

        <arg name="maxAcceleration" value="$(arg maxAcceleration)"/>
        <arg name="maxDeceleration" value="$(arg maxDeceleration)"/>

        <arg name="velocitySource" value="$(arg velocitySource)"/>
    </include>

    <!-- op_trajectory_generator -->
    <include file="$(find op_local_planner)/launch/op_trajectory_generator.launch">
    </include>

    <!-- op_motion_predictor -->
    <include file="$(find op_local_planner)/launch/op_motion_predictor.launch">
    </include>

    <!-- op_trajectory_evaluator -->
    <include file="$(find op_local_planner)/launch/op_trajectory_evaluator.launch">
        <arg name="enablePrediction" value="/adeye/motion_planning/op_trajectory_evaluator/enablePrediction"/>
        <arg name="horizontalSafetyDistance" value="/adeye/motion_planning/op_trajectory_evaluator/horizontalSafetyDistance"/>
        <arg name="verticalSafetyDistance" value="/adeye/motion_planning/op_trajectory_evaluator/verticalSafetyDistance"/>
    </include>

    <!-- op_behavior_selector -->
    <include file="$(find op_local_planner)/launch/op_behavior_selector.launch">
        <param name="evidence_tust_number" value="/adeye/motion_planning/op_behavior_selector/evidence_tust_number"/>
    </include>

    <!-- pure_pursuit config -->
    <node pkg="rostopic" type="rostopic" name="config_waypoint_follower_rostopic"
          args="pub -l /config/waypoint_follower autoware_config_msgs/ConfigWaypointFollower
        '{ header: auto, param_flag: /adeye/motion_planning/pure_pursuit_config/param_flag, velocity: /adeye/motion_planning/pure_pursuit_config/velocity, lookahead_distance: /adeye/motion_planning/pure_pursuit_config/lookahead_distance, lookahead_ratio: /adeye/motion_planning/pure_pursuit_config/lookahead_ratio, minimum_lookahead_distance: /adeye/motion_planning/pure_pursuit_config/minimum_lookahead_distance, displacement_threshold: /adeye/motion_planning/pure_pursuit_config/displacement_threshold, relative_angle_threshold: /adeye/motion_planning/pure_pursuit_config/relative_angle_threshold }' "
    />

    <!-- rosrun waypoint_follower pure_pursuit -->
    <node pkg="waypoint_follower" type="pure_pursuit" name="pure_pursuit" output="log" respawn="true">
        <param name="is_linear_interpolation" value="$(arg is_linear_interpolation)"/>
        <param name="publishes_for_steering_robot" value="$(arg publishes_for_steering_robot)"/>
    </node>

    <!-- twist_filter -->
    <include file="$(find waypoint_follower)/launch/twist_filter.launch"/>

</launch>
